metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -1156,9 +1169,32 @@ private Errors validateAlterPartitionData(
             }
         }
 
+        List<BrokerEpochMismatchReplica> brokerEpochMismatchReplicas =
+            brokerEpochMismatchReplicasForIsr(partitionData.newIsrWithEpochs());
+        if (!brokerEpochMismatchReplicas.isEmpty()) {
+            log.info("Rejecting AlterPartition request from node {} for {}-{} because " +
+                    "it specified some invalid ISR members {} that the broker epochs mismatch.",
+                    brokerId, topic.name, partitionId, brokerEpochMismatchReplicas);
+
+            return INELIGIBLE_REPLICA;
+        }
+
         return Errors.NONE;
     }
 
+    private List<BrokerEpochMismatchReplica> brokerEpochMismatchReplicasForIsr(List<BrokerState> brokerStates) {
```

Comment:
I wonder if we could combine this with ineligibleReplicasForIsr. Did you consider/try this?